Add support for SAML authentication.

ray-lee opened the pull request:
## References

* Fixes #9438
* Related to DSpace/dspace-angular#2937

## Description

This adds support for SAML authentication.

## Instructions for Reviewers

List of changes in this PR:
* Add VS Code project files to .gitignore, so I can use it for development.
* Amend the ConfigurationService API to allow working with configuration properties as hierarchies. This makes it much easier to deal with deeply nested properties required to fully configure SAML relying parties.
* Implement a SAML relying party service at `/server/saml2`.
* Implement a SAML authentication endpoint at `/server/api/authn/saml`.

See #9438 for details about the SAML relying party service and authentication endpoint.

This needs to be tested in conjunction with the UI changes in DSpace/dspace-angular#2937.

To test this, you'll need a SAML IdP. I've been developing against a personal auth0 instance, which you're welcome to use. 

This is the minimal configuration that you'll need to add to your local.cfg to use my auth0 instance as the IdP:
```
saml-relying-party.auth0.asserting-party.metadata-uri = https://dev-vynkcnqhac3c0s10.us.auth0.com/samlp/metadata/Vn8jWX0iFHtepmXi7rjZa9h5M1kqXNWY

saml-relying-party.auth0.attributes = \
  http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress => org.dspace.saml.EMAIL, \
  http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname => org.dspace.saml.GIVEN_NAME, \
  http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname => org.dspace.saml.SURNAME
```

Also enable SAML authentication, and set it to use the above relying party:
```
plugin.sequence.org.dspace.authenticate.AuthenticationMethod = org.dspace.authenticate.PasswordAuthentication,org.dspace.authenticate.SamlAuthentication

authentication-saml.relying-party-id = auth0
```

> [!IMPORTANT] 
> If you use this auth0 IdP, and your DSpace is running locally at http://localhost:8080: Following a successful login at the  IdP, the IdP will return HTML containing a form that will automatically be posted by the user's browser back to DSpace. This is how DSpace receives the identity assertion from the IdP. Since the auth0 IdP is at an https URL, and the form is being posted to an http URL, the browser will set the Origin header to `null` in order to avoid leaking information from a secure URL to an insecure one. To deal with this, you'll need to add `null` to the DSpace CORS allowed origins — _for this testing scenario only_:
> ```
> rest.cors.allowed-origins = ${dspace.ui.url},null
> ```
> (In production use, this effectively means that in order to use SAML authentication, DSpace must be installed at an https URL, and the URL of the IdP would need to be added to `rest.cors.allowed-origins`.)

To use a different IdP, see [the configuration example](https://github.com/ray-lee/DSpace/blob/saml/dspace/config/modules/saml-relying-party.cfg) for details about the available configuration options for a relying party. Consult with the administrator of your IdP to determine how to make the connection. Typically, the process is something like:

1. The administrator of the IdP provides a URL to the metadata for the IdP.
1. Configure a new relying party in DSpace, setting `asserting-party.metadata-uri` to the IdP's metadata URL. 
1. If encryption and/or signing are desired/required by the IdP, configure private keys and certificates into the relying party.
1. Ask the administrator of the IdP to release the user's email, first name, and last name as attributes, and ask what the names of those attributes will be. Configure the relying party's `attributes` property, to map those attribute names to `org.dspace.saml.EMAIL`, `org.dspace.saml.GIVEN_NAME`, and `org.dspace.saml.SURNAME`, as in the above example. 
1. Start DSpace, and provide the relying party's metadata URL (`/server/saml2/service-provider-metadata/{id}`) to the administrator of the IdP, so they can use it to configure their side. Alternatively, retrieve the metadata, and provide it to the administrator of the IdP through another channel.
1. The administrator of the IdP says that the configuration is complete. Testing can commence.

### Testing

1. Run DSpace from this branch with a SAML relying party and SAML authentication configured, as described above. 
1. Run the UI from DSpace/dspace-angular#2937.

#### To test SAML login as an existing DSpace user:

1. Log in as an Administrator using email/password authentication.
1. Create a new user, noting the email address you use. (If using the auth0 service, an email will actually be sent to this address.)
1. Log out.
1. Log in again, this time clicking the "Log in with SAML" button.
1. The login screen from the IdP should appear. If using the auth0 service, this will have a "sign up" link that you can use to create a user. Use the email address of the user you just added to DSpace.
1. After registration/login with the IdP, you should be redirected back to DSpace, and should be logged in as the user you created in step 2.
1. Log out. This should log you out of DSpace, but not out of the IdP.
1. Log in again. This should log you into DSpace as the same user, without showing the login screen from the IdP, since the login with the IdP is still active.
1. Log out again.
1. Log out of the IdP. If using the auth0 service, go to https://dev-vynkcnqhac3c0s10.us.auth0.com/v2/logout. If using another IdP, visit its logout page.
1. Log in using SAML again. You should see the login screen from the IdP this time, since you logged out of the IdP.
1. Log out of both DSpace and the IdP.

#### To test SAML login as a new, auto-registered DSpace user:

1. If using the auth0 service: ~~You can't self-register a new auth0 user for this test, because they don't let you set a first and last name when registering, and a first and last name are required to auto-register a DSpace user. Contact me for the login information for an auth0 user that has the first and last name set.~~ When self-registering an auth0 account, they don't let you set a first and last name on the account, but that information is required to auto-register a DSpace user. Contact me, and I can set the first/last names on your auth0 account using its API. That way you can use the same user to test the auto-registration scenario.

   If using your own IdP, identify a user on the IdP that does not have an existing DSpace account, and that does have the email, first name, and last name set.
1. Log in to DSpace using SAML. When the log in screen from the IdP appears, use the credentials of the user from step 1.
1. You should be redirected back to DSpace, and the End User Agreement should appear.
1. Accept the End User Agreement.
1. You should be logged in as a newly created DSpace user, with the email address, first name, and last name from the IdP.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ ] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
  _Sorry, this is about as small as I can make it, and have it be functional._
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
  _I'm not sure if this is necessary for these changes._
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

Commit:
Add VS Code settings to .gitignore.

Commit:
Add methods for working with hierarchical configuraton properties.

This adds methods to ConfigurationService and DSpaceConfigurationService to work with configuration properties as hierarchies, exposing the functionality available in Apache Commons Configuration.

Commit:
Add dspace-saml2 module.

Commit:
Add SAML login filter.

Commit:
Add SAML integration test.

mark-cooper committed:
Log incoming saml attributes from provider.

mark-cooper committed:
Add dspace saml failure handler.

Commit:
Change version to 8.0.

Also resolve version conflict in micrometer-observation.

Commit:
Change versions to 9.0-SNAPSHOT.

kshepherd commented on dspace-api/src/main/java/org/dspace/authenticate/SamlAuthentication.java:
```diff
@@ -0,0 +1,750 @@
+    protected Map<String, String> metadataHeaderMap = null;
+
+    // Maximum length for ePerson fields.
+    protected final int NAME_MAX_SIZE = 64;
```
`EPerson` name (full = first + last, or email if null) is stored in metadata and so technically has a max size of `Length.LONG32` - the email address limit is 64 characters, though email length is not validated later on in the register method.
Is this check maybe because there are situations where the first and last names are null, the `getFullName()` method then defaults to email, so you need to validate against email max length? Or can we increase the maximum name length?

ray-lee replied:
I admit that I just cribbed this from ShibAuthentication, which I used as a model for SamlAuthentication. I wonder if there was a restriction in place when ShibAuthentication was implemented?

kshepherd commented on dspace-api/src/main/java/org/dspace/authenticate/SamlAuthentication.java:
```diff
@@ -0,0 +1,750 @@
+
+        // Turn off authorizations to create a new user
+
+        context.turnOffAuthorisationSystem();
```
The `context.turnOffAuthorisationSystem()` and `context.restoreAuthSystemState()` calls should be wrapped in a try / catch / finally block, so that even if an exception is thrown, we properly restore the current context auth state.

This is done properly in the `autoCreateEPersonMetadataField()` method in this class, we just need to fix up the other instances of auth bypass.

ray-lee replied:
Fixed in 8f4aacd.

kshepherd commented on dspace-api/src/main/java/org/dspace/authenticate/SamlAuthentication.java:
```diff
@@ -0,0 +1,750 @@
+            lastName = lastName.substring(0, NAME_MAX_SIZE);
+        }
+
+        context.turnOffAuthorisationSystem();
```
As per previous comment, this should be in try / catch / finally so we always restore context auth state.

ray-lee replied:
Fixed in 8f4aacd.

kshepherd commented on dspace-api/src/main/java/org/dspace/authenticate/SamlAuthentication.java:
```diff
@@ -0,0 +1,750 @@
+     * @return True if a valid metadata field, otherwise false.
+     * @throws SQLException if database error
+     */
+    protected synchronized boolean checkIfEpersonMetadataFieldExists(Context context, String metadataName)
```
The camelCase on this method name is a bit inconsistent compared to `registerNewEPerson` and EPerson class name etc but I see that's also the case with `ShibAuthentication` so, not a blocker or requirement to fix imo

ray-lee replied:
Fixed in 8f4aacd.

kshepherd reviewed (changes requested):
Thanks @ray-lee , looks great! I have tested this branch (and the frontend) using your auth0 service for existing user login and it is working as expected. Could you please share the details for an auto-registration test as per your instructions? kim at shepherd dot nz. Then I'll try that out too.

I also did a code review, spending most time in the authentication plugin as that is the area I know best - I'm not an expert in SAML itself so I don't have as much authority on the saml2.* classes, but they look fine to me.

The auth plugin is great, very well documented, and I just noted a small question about max name length checks, and some improvements to make to context auth bypass.

**For other reviewers**: I'm thinking it'd be good for someone with access to another SAML IdP (or motivation to set one up at a test service) to try this in practice so we are not only testing against the same dev auth0 stack

Commit:
Wrap turnOffAuthorisationSystem calls in try/catch.

Also change case of checkIfEpersonMetadataFieldExists to checkIfEPersonMetadataFieldExists for consistency.

Commit:
Remove length validation.

mark-cooper commented:
I've tested this (and related) code and found it to work as described.

For testing it initially I used a local Docker SAML IDP, then a remote test SAML IDP (like Auth0).

I've also deployed a backported 7.6.1/2 branch to use for a number of production sites and they've been working successfully for at least a few months now.

So this LGTM 👍

misilot commented:
@mark-cooper can you share those back ports for 7.6.x?

mark-cooper commented:
@misilot sure, you can get them from @ray-lee 

https://github.com/ray-lee/DSpace/tree/saml-7
https://github.com/ray-lee/dspace-angular/tree/saml-7

Commit:
Bump Spring Boot version.

misilot commented:
Whoops, I commented on #9438 but I tested this against our Shibboleth IdP, and it works great! I'm excited to see this get included and having SAML as a native authentication method. 

Editorial note: I tested on DSpace 7.6.2

kshepherd commented:
rerunning the failed linux image build job (ant download issue)

tdonohue commented:
@ray-lee or @mark-cooper : Is there any way to get this PR updated?  It has minor merge conflicts.  I'd like to see if we can consider this for the 9.0 release. But to do so, we'll obviously need to get those merge conflicts resolved.  Thanks!

mark-cooper commented:
@tdonohue I'm meeting with @ray-lee Tues next week, and I believe we can likely have an update soon after (if Ray doesn't happen to get to it sooner) 👍

ray-lee commented:
@tdonohue I'm having trouble resolving the conflict, and I'm hoping someone can help, as this is reproducible on main.

This commit upgraded Spring Security from 6.3.4 to 6.4.0: https://github.com/DSpace/DSpace/commit/ec73c37554e21de50d8775bacd06d93aace751bf

As the comment on that line states, the version of Spring Security needs to be in sync with the version of Spring Boot. Spring Boot was not upgraded with Spring Security, and stayed at 3.3.5, which is not compatible with Spring Security 6.4.0. In order to sync up Spring Boot and Spring Security (such that Spring Security remains >= 6.4.0), Spring Boot needs to be upgraded to 3.4.0, and Spring Security to 6.4.1.

This causes a version conflict in snakeyaml, which I assume can be resolved by excluding the dependency on the older version of snakeyaml. Once that's done, the project builds, but there is an error at startup:

> Caused by: java.lang.ClassNotFoundException: org.eclipse.jetty.io.ClientConnector

I'm not sure where that is coming from. Is someone else able to take a look?

ray-lee commented:
I created an issue for the Spring versioning problem: https://github.com/DSpace/DSpace/issues/10203

Commit:
Merge branch 'main' into saml-merge-main

Commit:
Fix version conflict in velocity between dspace-saml2 and dspace-api.

Commit:
Update test for newer version of Spring Security.

ray-lee commented:
@tdonohue @mark-cooper the merge conflict has been resolved. I'm not sure if that Docker build failure is anything I did?

tdonohue commented:
@ray-lee : It appears (to me) that the Docker build process failure is specific to this PR.  It seems like Docker is failing to find the new `dspace-saml2` Maven module you have in this PR.  The maven process that Docker is using is not compiling that module & then throws an error when it doesn't exist.  Here's the final summary of that error, which occurs when building `dspace-server-webapp`:
```
#34 21.48 [INFO] ------------------< org.dspace:dspace-server-webapp >-------------------
#34 21.48 [INFO] Building DSpace Server Webapp 9.0-SNAPSHOT                       [12/12]
#34 21.48 [INFO]   from dspace-server-webapp/pom.xml
#34 21.48 [INFO] --------------------------------[ jar ]---------------------------------
#34 22.09 [WARNING] The POM for org.dspace:dspace-saml2:jar:9.0-SNAPSHOT is missing, no dependency information available
#34 22.55 [INFO] ------------------------------------------------------------------------
#34 22.55 [INFO] Reactor Summary for DSpace Parent Project 9.0-SNAPSHOT:
#34 22.55 [INFO] 
#34 22.55 [INFO] DSpace Parent Project .............................. SUCCESS [  5.187 s]
#34 22.55 [INFO] DSpace Services Framework :: API and Implementation  SUCCESS [  1.723 s]
#34 22.55 [INFO] DSpace Kernel :: API and Implementation ............ SUCCESS [ 11.004 s]
#34 22.55 [INFO] DSpace Addon Modules ............................... SUCCESS [  0.013 s]
#34 22.55 [INFO] DSpace Kernel :: Additions and Local Customizations  SUCCESS [  0.252 s]
#34 22.55 [INFO] DSpace OAI-PMH ..................................... SUCCESS [  0.578 s]
#34 22.55 [INFO] DSpace Assembly and Configuration .................. SUCCESS [  0.026 s]
#34 22.55 [INFO] DSpace RDF ......................................... SUCCESS [  0.066 s]
#34 22.56 [INFO] DSpace IIIF ........................................ SUCCESS [  0.696 s]
#34 22.56 [INFO] DSpace SWORD ....................................... SUCCESS [  0.119 s]
#34 22.56 [INFO] DSpace SWORD v2 .................................... SUCCESS [  0.322 s]
#34 22.56 [INFO] DSpace Server Webapp ............................... FAILURE [  1.079 s]
#34 22.56 [INFO] ------------------------------------------------------------------------
#34 22.56 [INFO] BUILD FAILURE
#34 22.56 [INFO] ------------------------------------------------------------------------
#34 22.56 [INFO] Total time:  21.290 s
#34 22.56 [INFO] Finished at: 2025-01-17T23:05:32Z
#34 22.56 [INFO] ------------------------------------------------------------------------
#34 22.56 [ERROR] Failed to execute goal on project dspace-server-webapp: Could not resolve dependencies for project org.dspace:dspace-server-webapp:jar:9.0-SNAPSHOT
#34 22.56 [ERROR] dependency: org.dspace:dspace-saml2:jar:9.0-SNAPSHOT (compile)
#34 22.56 [ERROR] 	Could not find artifact org.dspace:dspace-saml2:jar:9.0-SNAPSHOT in maven-snapshots (https://oss.sonatype.org/content/repositories/snapshots)
```

Commit:
Add dspace-saml2 to Docker dependencies.

ray-lee commented:
@tdonohue All set now. I figured out I needed to add dspace-saml2 to the new Dockerfile.dependencies file.

reineruhry commented:
I successfully tested this branch by following the steps below:

1. Pulled the DSpace Backend and Frontend branches.
2. Installed all required Frontend dependencies.
3. Configured the local.cfg file with the recommended settings for Auth0.
4. Built the Docker containers using the following command:
    docker-compose -f docker-compose.yml -f docker-compose-cli.yml build
5. Started DSpace with:
    docker-compose -p d8 up -d
6. Created an admin account on DSpace.
7. Started the frontend using:
    npm start
8. Logged in as Admin and created a user account.
9. Used Chrome (a different browser) to log in via SAML authentication.

For the second test, I updated the local.cfg file to add a second identity provider (dummy local), and successfully logged in again.

I did encounter some CORS issues in Safari, specifically with the use of the localhost address. However, these issues are unrelated to this pull request.

tdonohue commented on dspace/config/modules/authentication-saml.cfg:
```diff
@@ -0,0 +1,41 @@
+#   org.dspace.saml.PHONE => phone
+
+# If the ePerson metadata field is not found, should it be created automatically?
+authentication-saml.eperson.metadata.autocreate = true
```
Similar to above, this should be set to "false" by default.  We should also note in the comment that the metadata fields will be created in the existing "eperson" schema.

Just a general comment.  Any of the configurations in this file that have a defined "default value" in Java code can be commented out by default.   I believe that means that most of these configurations can be commented out by default.  (This would be more similar to other existing `authentication-*.cfg` files, where many of the configs are commented out by default.)

ray-lee replied:
Fixed in 481cdd1. I also added defaults for some settings that didn't have them, but should have (in a919719).

tdonohue reviewed (changes requested):
@ray-lee : Thanks for this work!  I gave this a code review today, and it mostly looks good to me.  I do have a few minor inline suggestions though...the biggest one is that I think we should *disable* the autocreation of metadata fields by default.  Beyond that, everything is very minor.

I haven't had a chance to test this myself yet, so this was just a code review.  But, it's great to see that others are already testing it successfully.

tdonohue reviewed (commented):
@ray-lee :  I had a chance to test this out today after all.  I found it all worked well using your test IdP and the "Instructions for Reviewers" in the PR description. I ran into no bugs or issues.  So, my only feedback is the [code review I left above](https://github.com/DSpace/DSpace/pull/9470#pullrequestreview-2651783276).  Beyond that, this looks good to me. Thanks again!

kshepherd reviewed (approved):
since the testing issue lately seemed to be my end, it has worked for me before, and it works for Tim, I am +1 now! (pending tim's requested small changes). thanks @ray-lee !

Commit:
Change default of authentication-saml.eperson.metadata.autocreate to false.

Commit:
Add default values for configuration settings.

Commit:
Comment out saml configuration settings that have defaults.

Commit:
Correct constant name and comment.

Commit:
Remove redundant metadata name check.

Commit:
Fix checkstyle errors.

ray-lee commented on dspace-api/src/main/java/org/dspace/authenticate/SamlAuthentication.java:
```diff
@@ -0,0 +1,715 @@
+        String[] mappingString = configurationService.getArrayProperty("authentication-saml.eperson.metadata");
+
+        boolean autoCreate = configurationService
+            .getBooleanProperty("authentication-saml.eperson.metadata.autocreate", true);
```
Fixed in b4a2df4.

ray-lee commented on dspace-api/src/main/java/org/dspace/authenticate/SamlAuthentication.java:
```diff
@@ -0,0 +1,715 @@
+    /**
+     * Validate Postgres Column Names
+     */
+    protected final String COLUMN_NAME_REGEX = "^[_A-Za-z0-9]+$";
```
Fixed in 5b41f55.

ray-lee commented on dspace-api/src/main/java/org/dspace/authenticate/SamlAuthentication.java:
```diff
@@ -0,0 +1,715 @@
+        }
+
+        // The phone is a predefined field
+        if ("phone".equals(metadataName)) {
```
Fixed in b40b047.

ray-lee commented:
@tdonohue I've made the requested changes.

tdonohue reviewed (approved):
👍 Thanks @ray-lee !  This now looks good to me!  As this is now at +2 (plus a few other positive testers), I'll merge shortly so this will be included in the upcoming DSpace 9.0 release.

tdonohue commented:
I forgot to mention. We need to add documentation for SAML authentication to the 9.0 Wiki Documentation.  However, that 9.0 Wiki Space is not yet created (and I won't be able to create it today as the wiki is having some issues that should hopefully be resolved by an upgrade tonight).

Flagging this as "needs documentation" until appropriate documentation can be added similar to other authentication plugins https://wiki.lyrasis.org/display/DSDOC8x/Authentication+Plugins 

(I'll try to remember to update this comment once our DSpace 9.0 Documentation space is created)

tdonohue commented:
NOTE: Our 9.0 documentation wiki page now exists.  This feature needs to have documentation added to that space for the 9.0 release: https://wiki.lyrasis.org/display/DSDOC9x

If you have questions about where to add the documentation, please get in touch.  (This is a generic comment that I'm adding to many PRs at once)

tdonohue commented:
@mark-cooper or @ray-lee : Will one of you be able to provide documentation for this feature in our 9.0 docs? https://wiki.lyrasis.org/display/DSDOC9x   Ideally, basic documentation would exist prior to our [9.0 Testathon](https://wiki.lyrasis.org/display/DSPACE/DSpace+Release+9.0+Testathon+Page) (which starts on Monday, April 7).  But, if it cannot exist by then, it should hopefully be created shortly thereafter.

tdonohue commented:
@mark-cooper or @ray-lee : Reminder that we are still requiring *Documentation* for SAML Authentication for DSpace 9.0.  The docs should be added to https://wiki.lyrasis.org/display/DSDOC9x/Authentication+Plugins

tdonohue commented:
NOTE: This feature has now been documented at https://wiki.lyrasis.org/display/DSDOC9x/Authentication+Plugins#AuthenticationPlugins-SAML(SecurityAssertionMarkupLanguage)Authentication
